inlanecruising_plugin/src/calculation.cpp

```diff
@@ -0,0 +1,308 @@
+#include <algorithm>
+#include <tf/transform_datatypes.h>
+#include "inlanecruising_plugin.h"
+#include <tuple>
+#include <cav_msgs/ManeuverPlan.h>
+#include <cav_msgs/LaneFollowingManeuver.h>
+#include <carma_wm/WorldModel.h>
+#include <carma_wm/Geometry.h>
+#include <limits>
+#include <unordered_set>
+
+/*
+#
+# 2d x-axis position of the vehicle center of gravity in meters
+# This position is in a fixed inertial frame which vehicle motion is described in
+#
+float64 X_pos_global
+#
+# 2d y-axis position of the vehicle center of gravity in meters
+# This position is in a fixed inertial frame which vehicle motion is described in
+#
+float64 Y_pos_global
+#
+# The orientation of the vehicle's longitudinal axis in radians
+# This orientation is in a fixed inertial frame which vehicle motion is described in
+#
+float64 orientation
+#
+# longitudinal velocity of the vehicle center of gravity in ms in its body frame
+#
+float64 longitudinal_vel
+
+*/
+
+namespace inlanecruising_plugin
+{
+
+std::vector<PointSpeedPair> points_in_time_boundary(const std::vector<PointSpeedPair>& points, double time_span) {
+  std::vector<PointSpeedPair> output;
+  output.reserve(points.size());
+  double current_time = 0;
+  PointSpeedPair prev_pair;
+  bool first = true;
+  for (const auto& p : points) {
+    if (current_time > time_span) {
+      return output;
+    }
+
+   ROS_WARN_STREAM("Evaluating time of point: " << std::get<0>(p).x() << ", " << std::get<0>(p).y());
+
+
+    if (first) {
+      output.push_back(p);
+      prev_pair = p;
+      first = false;
+      continue;
+    }
+
+    double delta_d = lanelet::geometry::distance2d(std::get<0>(p), std::get<0>(prev_pair));
+
+    current_time += delta_d / std::get<1>(prev_pair);
+
+    output.push_back(p);
+    prev_pair = p;
+  }
+
+  return output;
+}
+
+std::vector<PointSpeedPair> maneuvers_to_points(const std::vector<cav_msgs::Maneuver>& maneuvers,
+                                                double max_starting_downtrack, const carma_wm::WorldModelConstPtr& wm)
+{
+  std::vector<PointSpeedPair> points_and_target_speeds;
+  std::unordered_set<lanelet::Id> visited_lanelets;
+
+  bool first = true;
+  ROS_WARN_STREAM("VehDowntrack: " << max_starting_downtrack);
+  for (const auto& manuever : maneuvers)
+  {
+    if (manuever.type != cav_msgs::Maneuver::LANE_FOLLOWING)
+    {
+      throw std::invalid_argument("In-Lane Cruising does not support this maneuver type");
+    }
+
+    cav_msgs::LaneFollowingManeuver lane_following_maneuver = manuever.lane_following_maneuver;
+
+    double starting_downtrack = lane_following_maneuver.start_dist;
+    if (first) {
+      if (starting_downtrack > max_starting_downtrack) {
+        starting_downtrack = max_starting_downtrack;
+      }
+      first = false;
+    }
+
+    ROS_WARN_STREAM("Used downtrack: " << starting_downtrack);
+
+    auto lanelets = wm->getLaneletsBetween(starting_downtrack, lane_following_maneuver.end_dist, true);
```

Comment:
Hmm, this is kinda complicated since we're working in distance domain here, and arbitrator really cares about total plan duration in seconds. I'm wondering if we need to just definitively pick one to use, since the transform between these frames is related to speed, which is what we're also changing/computing. Maybe it doesn't matter here, since we're at the tactical level, but it's kind of got my head spinning.

Reply:
I agree I think we need to hash this out a bit more. In this case, I was grabbing the related lanelets and then doing the time constraint on a future step.